File: iotdb-client/session/src/main/java/org/apache/iotdb/session/NodesSupplier.java
package org.apache.iotdb.session;

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import org.apache.iotdb.common.rpc.thrift.TEndPoint;
import org.apache.iotdb.isession.INodeSupplier;
import org.apache.iotdb.isession.SessionDataSet;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.ZoneId;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class NodesSupplier implements INodeSupplier, Runnable {

  private static final Logger LOGGER = LoggerFactory.getLogger(NodesSupplier.class);

  private static final long UPDATE_PERIOD_IN_S = 60;
  private static final String SHOW_DATA_NODES_COMMAND = "SHOW DATANODES";

  private static final String STATUS_COLUMN_NAME = "Status";

  private static final String IP_COLUMN_NAME = "RpcAddress";

  private static final String PORT_COLUMN_NAME = "RpcPort";

  private static final String REMOVING_STATUS = "Removing";

  // it's ok that TIMEOUT_IN_MS is larger than UPDATE_PERIOD_IN_S, because the next update request
  // won't be scheduled until last time is done.
  private static final long TIMEOUT_IN_MS = 60_000;

  private static final int FETCH_SIZE = 10_000;

  // availableNodes won't be updated frequently, so we use CopyOnWriteArrayList which is thread-safe
  // and is optimized for scenarios of reading more and writing less
  @SuppressWarnings("java:S3077")
  private volatile List<TEndPoint> availableNodes = new CopyOnWriteArrayList<>();

  private final boolean useSSL;
  private final String trustStore;
  private final String trustStorePwd;
  private final boolean enableRPCCompression;
  private final String userName;

  private final String password;

  private final ZoneId zoneId;

  private final int thriftDefaultBufferSize;

  private final int thriftMaxFrameSize;

  private final int connectionTimeoutInMs;

  private final String version;

  private final QueryEndPointPolicy policy = new RoundRobinPolicy();

  private ThriftConnection client;

  private volatile boolean closed = false;

  @SuppressWarnings("unsafeThreadSchedule")
  public static NodesSupplier createNodeSupplier(
      List<TEndPoint> endPointList,
      ScheduledExecutorService executorService,
      String userName,
      String password,
      ZoneId zoneId,
      int thriftDefaultBufferSize,
      int thriftMaxFrameSize,
      int connectionTimeoutInMs,
      boolean useSSL,
      String trustStore,
      String trustStorePwd,
      boolean enableRPCCompression,
      String version) {

    NodesSupplier nodesSupplier =
        new NodesSupplier(
            endPointList,
            userName,
            password,
            zoneId,
            thriftDefaultBufferSize,
            thriftMaxFrameSize,
            connectionTimeoutInMs,
            useSSL,
            trustStore,
            trustStorePwd,
            enableRPCCompression,
            version);

    // call executorService.scheduleAtFixedRate here in a separate line
    // instead of the constructor to prevent leaking the "this" reference to
    // another thread, which will cause unsafe publication of this instance.
    executorService.scheduleAtFixedRate(nodesSupplier, 0, UPDATE_PERIOD_IN_S, TimeUnit.SECONDS);

    return nodesSupplier;
  }

  private NodesSupplier(
      List<TEndPoint> endPointList,
      String userName,
      String password,
      ZoneId zoneId,
      int thriftDefaultBufferSize,
      int thriftMaxFrameSize,
      int connectionTimeoutInMs,
      boolean useSSL,
      String trustStore,
      String trustStorePwd,
      boolean enableRPCCompression,
      String version) {
    this.availableNodes.addAll(new HashSet<>(endPointList));
    this.userName = userName;
    this.password = password;
    this.useSSL = useSSL;
    this.trustStore = trustStore;
    this.trustStorePwd = trustStorePwd;
    this.enableRPCCompression = enableRPCCompression;
    this.zoneId = zoneId == null ? ZoneId.systemDefault() : zoneId;
    this.thriftDefaultBufferSize = thriftDefaultBufferSize;
    this.thriftMaxFrameSize = thriftMaxFrameSize;
    this.connectionTimeoutInMs = connectionTimeoutInMs;
    this.version = version;
  }

  // the method will only be called while reconnecting, so it's ok to copy set each time
  // and the List needn't be thread-safe, because it will only be used in one thread.
  @Override
  public List<TEndPoint> get() {
    return availableNodes;
  }

  @Override
  public void run() {
    if (closed) {
      if (client != null) {
        destroyCurrentClient();
      }
      return;
    }
    if (client == null) {
      for (TEndPoint endPoint : availableNodes) {
        if (createConnection(endPoint)) {
          break;
        }
      }
    }

    // make sure the following code block can run thread-safely with close() method.
    synchronized (this) {
      if (client != null && !updateDataNodeList()) {
        destroyCurrentClient();
      }
    }
  }

  private boolean createConnection(TEndPoint endPoint) {
    client =
        new ThriftConnection(
            endPoint, thriftDefaultBufferSize, thriftMaxFrameSize, connectionTimeoutInMs);
    try {
      client.init(
          useSSL,
          trustStore,
          trustStorePwd,
          userName,
          password,
          enableRPCCompression,
          zoneId,
          version);
      return true;
    } catch (Exception e) {
      LOGGER.warn("Failed to create connection with {}.", endPoint);
      destroyCurrentClient();
      return false;
    }
  }

  private synchronized void destroyCurrentClient() {
    if (client != null) {
      client.close();
      client = null;
    }
  }

  @Override
  public void close() {
    closed = true;
    destroyCurrentClient();
  }

  @Override
  public Optional<TEndPoint> getQueryEndPoint() {
    if (availableNodes == null || availableNodes.isEmpty()) {
      return Optional.empty();
    } else {
      return Optional.of(policy.chooseOne(get()));
    }
  }

  private boolean updateDataNodeList() {
    try (SessionDataSet sessionDataSet =
        client.executeQueryStatement(SHOW_DATA_NODES_COMMAND, TIMEOUT_IN_MS, FETCH_SIZE)) {
      SessionDataSet.DataIterator iterator = sessionDataSet.iterator();
      List<TEndPoint> res = new ArrayList<>();
      while (iterator.next()) {
        String ip = iterator.getString(IP_COLUMN_NAME);
        // ignore 0.0.0.0 and removing DN
        if (!REMOVING_STATUS.equals(iterator.getString(STATUS_COLUMN_NAME))
            && !"0.0.0.0".equals(ip)) {
          String port = iterator.getString(PORT_COLUMN_NAME);
          if (ip != null && port != null) {
            res.add(new TEndPoint(ip, Integer.parseInt(port)));
          }
        }
      }
      // replace the older ones.
      if (!res.isEmpty()) {
        availableNodes = res;
      }
      return true;
    } catch (Exception e) {
      LOGGER.warn("Failed to fetch data node list from {}.", client.endPoint);
      return false;
    }
  }
}

File: iotdb-client/client-cpp/src/main/NodesSupplier.cpp
/**
* Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "NodesSupplier.h"
#include "Session.h"
#include "SessionDataSet.h"
#include <algorithm>
#include <iostream>
#include <utility>

const std::string NodesSupplier::SHOW_DATA_NODES_COMMAND = "SHOW DATANODES";
const std::string NodesSupplier::RUNNING_STATUS = "Running";
const std::string NodesSupplier::STATUS_COLUMN_NAME = "Status";
const std::string NodesSupplier::IP_COLUMN_NAME = "RpcAddress";
const std::string NodesSupplier::PORT_COLUMN_NAME = "RpcPort";
const std::string NodesSupplier::REMOVING_STATUS = "Removing";

const int64_t NodesSupplier::TIMEOUT_IN_MS = 60000;
const int NodesSupplier::FETCH_SIZE = 10000;
const int NodesSupplier::THRIFT_DEFAULT_BUFFER_SIZE = 4096;
const int NodesSupplier::THRIFT_MAX_FRAME_SIZE = 1048576;
const int NodesSupplier::CONNECTION_TIMEOUT_IN_MS = 1000;

TEndPoint RoundRobinPolicy::select(const std::vector<TEndPoint>& nodes) {
    static std::atomic_uint index{0};

    if (nodes.empty()) {
        throw IoTDBException("No available nodes");
    }

    return nodes[index++ % nodes.size()];
}

StaticNodesSupplier::StaticNodesSupplier(const std::vector<TEndPoint>& nodes,
                                       NodeSelectionPolicy policy)
    : availableNodes_(nodes), policy_(std::move(policy)) {}

boost::optional<TEndPoint> StaticNodesSupplier::getQueryEndPoint() {
    try {
        if (availableNodes_.empty()) {
            return boost::none;
        }
        return policy_(availableNodes_);
    } catch (const IoTDBException& e) {
        return boost::none;
    }
}

std::vector<TEndPoint> StaticNodesSupplier::getEndPointList() {
    return availableNodes_;
}

StaticNodesSupplier::~StaticNodesSupplier() = default;

std::shared_ptr<NodesSupplier> NodesSupplier::create(
    std::vector<TEndPoint> endpoints,
    std::string userName, std::string password, std::string zoneId,
    int32_t thriftDefaultBufferSize, int32_t thriftMaxFrameSize,
    int32_t connectionTimeoutInMs, bool useSSL, bool enableRPCCompression,
    std::string version, std::chrono::milliseconds refreshInterval,
    NodeSelectionPolicy policy) {
    if (endpoints.empty()) {
        return nullptr;
    }
    auto supplier = std::make_shared<NodesSupplier>(
        userName, password, zoneId, thriftDefaultBufferSize,
        thriftMaxFrameSize, connectionTimeoutInMs, useSSL,
        enableRPCCompression, version, std::move(endpoints), std::move(policy)
    );
    supplier->startBackgroundRefresh(refreshInterval);
    return supplier;
}

NodesSupplier::NodesSupplier(
    std::string userName, std::string password, const std::string& zoneId,
    int32_t thriftDefaultBufferSize, int32_t thriftMaxFrameSize,
    int32_t connectionTimeoutInMs, bool useSSL, bool enableRPCCompression,
    std::string version, std::vector<TEndPoint> endpoints, NodeSelectionPolicy policy) : userName_(std::move(userName)), password_(std::move(password)), zoneId_(zoneId),
    thriftDefaultBufferSize_(thriftDefaultBufferSize), thriftMaxFrameSize_(thriftMaxFrameSize),
    connectionTimeoutInMs_(connectionTimeoutInMs), useSSL_(useSSL), enableRPCCompression_(enableRPCCompression), version(version), endpoints_(std::move(endpoints)),
    selectionPolicy_(std::move(policy)) {
    deduplicateEndpoints();
}

std::vector<TEndPoint> NodesSupplier::getEndPointList() {
    std::lock_guard<std::mutex> lock(mutex_);
    return endpoints_;
}

TEndPoint NodesSupplier::selectQueryEndpoint() {
    std::lock_guard<std::mutex> lock(mutex_);
    try {
        return selectionPolicy_(endpoints_);
    } catch (const std::exception& e) {
        log_error("NodesSupplier::selectQueryEndpoint exception: %s", e.what());
        throw IoTDBException("NodesSupplier::selectQueryEndpoint exception, " + std::string(e.what()));
    }
}

boost::optional<TEndPoint> NodesSupplier::getQueryEndPoint() {
    try {
        return selectQueryEndpoint();
    } catch (const IoTDBException& e) {
        return boost::none;
    }
}

NodesSupplier::~NodesSupplier() {
    stopBackgroundRefresh();
    if (client_ != nullptr) {
        client_->close();
    }
}

void NodesSupplier::deduplicateEndpoints() {
    std::vector<TEndPoint> uniqueEndpoints;
    uniqueEndpoints.reserve(endpoints_.size());
    for (const auto& endpoint : endpoints_) {
        if (std::find(uniqueEndpoints.begin(), uniqueEndpoints.end(), endpoint) == uniqueEndpoints.end()) {
            uniqueEndpoints.push_back(endpoint);
        }
    }
    endpoints_ = std::move(uniqueEndpoints);
}

void NodesSupplier::startBackgroundRefresh(std::chrono::milliseconds interval) {
    isRunning_ = true;
    refreshEndpointList();
    refreshThread_ = std::thread([this, interval] {
        while (isRunning_) {
            refreshEndpointList();
            std::unique_lock<std::mutex> cvLock(this->mutex_);
            refreshCondition_.wait_for(cvLock, interval, [this]() {
                return !isRunning_.load();
            });
        }
    });
}

std::vector<TEndPoint> NodesSupplier::fetchLatestEndpoints() {
  for (const auto& endpoint : endpoints_) {
    try {
      if (client_ == nullptr) {
        client_ = std::make_shared<ThriftConnection>(endpoint);
        client_->init(userName_, password_, enableRPCCompression_, zoneId_, version);
      }

      auto sessionDataSet = client_->executeQueryStatement(SHOW_DATA_NODES_COMMAND);

      uint32_t columnAddrIdx = -1, columnPortIdx = -1, columnStatusIdx = -1;
      auto columnNames = sessionDataSet->getColumnNames();
      for (uint32_t i = 0; i < columnNames.size(); i++) {
        if (columnNames[i] == IP_COLUMN_NAME) {
          columnAddrIdx = i;
        } else if (columnNames[i] == PORT_COLUMN_NAME) {
          columnPortIdx = i;
        } else if (columnNames[i] == STATUS_COLUMN_NAME) {
          columnStatusIdx = i;
        }
      }

      if (columnAddrIdx == -1 || columnPortIdx == -1 || columnStatusIdx == -1) {
        throw IoTDBException("Required columns not found in query result.");
      }

      std::vector<TEndPoint> ret;
      while (sessionDataSet->hasNext()) {
        auto record = sessionDataSet->next();
        std::string ip;
        int32_t port = 0;
        std::string status;

        if (record->fields.at(columnAddrIdx).stringV.is_initialized()) {
          ip = record->fields.at(columnAddrIdx).stringV.value();
        }
        if (record->fields.at(columnPortIdx).intV.is_initialized()) {
          port = record->fields.at(columnPortIdx).intV.value();
        }
        if (record->fields.at(columnStatusIdx).stringV.is_initialized()) {
          status = record->fields.at(columnStatusIdx).stringV.value();
        }

        if (ip == "0.0.0.0" || status != RUNNING_STATUS) {
          log_warn("Skipping invalid node: " + ip + ":" + std::to_string(port));
          continue;
        }
        TEndPoint newEndpoint;
        newEndpoint.ip = ip;
        newEndpoint.port = port;
        ret.emplace_back(newEndpoint);
      }
      return ret;  // success
    } catch (const std::exception& e) {
      log_warn("Failed to fetch endpoints from " + endpoint.ip + ":" +
               std::to_string(endpoint.port) + " , error=" + e.what());
      client_.reset();  // reset client before retrying next endpoint
      continue;         // try next endpoint
    }
  }
  throw IoTDBException("NodesSupplier::fetchLatestEndpoints failed: all nodes unreachable.");
}

void NodesSupplier::refreshEndpointList() {
    try {
        auto newEndpoints = fetchLatestEndpoints();
        if (newEndpoints.empty()) {
            return;
        }

        std::lock_guard<std::mutex> lock(mutex_);
        endpoints_.swap(newEndpoints);
        deduplicateEndpoints();
    } catch (const IoTDBException& e) {
        log_error(std::string("NodesSupplier::refreshEndpointList failed: ") + e.what());
    }
}

void NodesSupplier::stopBackgroundRefresh() noexcept {
    if (isRunning_.exchange(false)) {
        refreshCondition_.notify_all();
        if (refreshThread_.joinable()) {
            refreshThread_.join();
        }
    }
}